python/revng/tupletree.py

```diff
@@ -205,21 +215,42 @@ def id(self):  # noqa: A003
 
     @classmethod
     def yaml_representer(cls, dumper: yaml.dumper.Dumper, instance: "Reference"):
-        return dumper.represent_str(repr(instance))
+        return dumper.represent_str(instance._ref_str)
 
     def __repr__(self):
+        if self._ref_str == "":
+            return "<Invalid Reference>"
         return self._ref_str
 
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return False
+        return self._ref_str == other._ref_str and self.referenced_obj == other.referenced_obj
+
+    def is_valid(self):
+        return self._ref_str != ""
+
 
 def init_reference_yaml_classes(_: Type[yaml.Loader], dumper: Type[yaml.Dumper]):
     dumper.add_representer(Reference, Reference.yaml_representer)
 
 
-def _field_is_optional(field):
+def _field_is_default(field, value):
     field_metadata = field.metadata
     if field_metadata is None:
         return False
-    return field_metadata.get("optional", False)
+    is_optional = field_metadata.get("optional", False)
+    if not is_optional:
+        return False
+    factory = field_metadata["default_value"]
+    return value == factory()
+
+
+def _field_check_invalid(value):
+    is_valid_fun = getattr(value, "is_valid", None)
+    if is_valid_fun is None or not inspect.isfunction(is_valid_fun):
+        return False
+    return not is_valid_fun()
```

Comment:
1. This is an extension of what was used previously, and works with native data types (e.g. `int` and `str`)
2. I am not following, where's the negation? It would be weirder to have `if _field_is_default(field, field_val) or not _field_check_valid(field_val)`

Reply:
1. There are very few scalar data types, I don't think this warrants using introspection. I'd avoid `getattr` as much as possible.
2. You're assuming this function will be used in a single place. If you don't assume this, which is a reasonable approach for a function, it's better to have `not`. Let's go with `not`.

Reply:
Re-checking, only `Reference` had a `is_valid` function, so I removed the function entirely and special-cased the `if`